zone.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 The Inet.Af AUTHORS. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !go1.19
+
+package netaddr
+
+import (
+	"runtime"
+	"sync"
+	"unsafe"
+)
+
+// zone is the IPv6 zone and its generation count to prevent the finalizer
+// from deleting weak references from the uniqZone map.
+type zone struct {
+	// name is the IPv6 zone.
+	// It is immutable.
+	name string
+
+	// gen is guarded by zmu and is incremented whenever this zone
+	// is returned.
+	gen int64
+}
+
+var (
+	z0    = (*zone)(nil)
+	z4    = new(zone)
+	z6noz = new(zone)
+)
+
+// zmu guards uniqZone, a weakref map of *zones by zoneName.
+var (
+	zmu      sync.Mutex
+	uniqZone = map[string]uintptr{} // zone name to its uintptr(*zone)
+)
+
+// internZone returns an interned zone for zoneName.
+//
+// We intern zones to guarantee that two zone pointers are equal
+// iff their names are equal, so that IPs can be compared, used as map keys, etc.
+//
+// Interning is simple if you don't require that unused zones be garbage collectable.
+// But we do require that; we don't want to be DOS vector.
+// We do this by using a uintptr to hide our zone pointer from the garbage collector,
+// and using a finalizer to eliminate our zone pointer when no other code is using it.
+// (The checkptr runtime check prevents exactly these kinds of shenanigans; we disable it.)
```

Comment:
Can you say more words as to what exactly checkptr enforces/detects, that we're disabling? At first glance I don't understand what exactly about these shenanigans make it angry.

Reply:
Quoting the 1.14 release notes, checkptr is "a compile-time option for adding instrumentation to check that Go code is following unsafe.Pointer safety rules dynamically."

Package unsafe says:

>  The remaining patterns enumerate the only valid conversions from uintptr to Pointer. 
>  (3) Conversion of a Pointer to a uintptr and back, with arithmetic. 
>  (4) Conversion of a Pointer to a uintptr when calling syscall.Syscall. 
>  (5) Conversion of the result of reflect.Value.Pointer or reflect.Value.UnsafeAddr from uintptr to Pointer. 
>  (6) Conversion of a reflect.SliceHeader or reflect.StringHeader Data field to or from Pointer. 

Storing a uintptr and later making it back into an unsafe.Pointer is the golden example of what you're not supposed to do with package unsafe.

Which (if any) of these words would be helpful to add to the comments here? They're mostly about checkptr and the unsafe rules, not about our code.